sql/org_mozilla_firefox_derived/event_types_v1/init.sql

```diff
@@ -0,0 +1,132 @@
+CREATE OR REPLACE TABLE
+  org_mozilla_firefox_derived.event_types_v1
+AS
+WITH sample AS (
+  SELECT
+    name AS event,
+    * EXCEPT (name)
+  FROM
+    org_mozilla_firefox.events
+  CROSS JOIN
+    UNNEST(events) AS event
+  WHERE
+    DATE(submission_timestamp) >= '2020-01-01'
+),
+primary_event_types AS (
+  SELECT
+    category,
+    event,
+    MIN(timestamp) AS first_timestamp,
+    ROW_NUMBER() OVER (ORDER BY MIN(timestamp) ASC, category ASC, event ASC) AS primary_index,
+  FROM
+    sample
+  GROUP BY
+    category,
+    event
+),
+event_property_indices AS (
+  SELECT
+    category,
+    event,
+    MIN(timestamp) AS first_timestamp,
+    event_property.key AS event_property,
+    ROW_NUMBER() OVER (
+      PARTITION BY
+        category,
+        event
+      ORDER BY
+        MIN(timestamp) ASC
+    ) AS event_property_index,
+  FROM
+    sample,
+    UNNEST(extra) AS event_property
+  GROUP BY
+    category,
+    event,
+    event_property
+),
+event_property_value_indices AS (
+  SELECT
+    category,
+    event,
+    MIN(timestamp) AS first_timestamp,
+    event_property.key AS event_property,
+    event_property.value AS event_property_value,
+    ROW_NUMBER() OVER (
+      PARTITION BY
+        category,
+        event,
+        event_property.key
+      ORDER BY
+        MIN(timestamp) ASC
+    ) AS event_property_value_index,
+  FROM
+    sample,
+    UNNEST(extra) AS event_property
+  GROUP BY
+    category,
+    event,
+    event_property,
+    event_property_value
+),
+per_event_property AS (
+  SELECT
+    category,
+    event,
+    event_property,
+    event_property_index,
+    ARRAY_AGG(
+      STRUCT(
+        event_property_value AS key,
+        udf.event_code_points_to_string([event_property_value_index]) AS value,
+        event_property_value_index AS index
+      )
+      ORDER BY
+        event_property_value_index ASC
+    ) AS values,
+  FROM
+    event_property_value_indices
+  INNER JOIN
+    event_property_indices
+  USING
+    (category, event, event_property)
+  GROUP BY
+    category,
+    event,
+    event_property,
+    event_property_index
+),
+per_event AS (
+  SELECT
+    category,
+    event,
+    first_timestamp,
+    primary_index AS numeric_index,
+    udf.event_code_points_to_string([primary_index]) AS index,
```

Comment:
A question that came to my mind about this model in general: What if there are so many different events that not all of them can be represented as valid code points because they exceed the maximum valid value? I guess it might be possible to split up the code point and have multiple represent one event?

Reply:
Right, there are limitations. One of them is that we are limited to ~1M event types, and ~1M values per event property. We have never really gotten close to that point, and in general I would direct people at that point to consolidate events and use event properties to differentiate between them (if they need to).